File: InsertCSV_data.js

var MongoClient = require('mongodb').MongoClient;
var async = require('async');
var url = "mongodb://localhost:27017";

const csv = require('csv-parser');
const fs = require('fs');

MongoClient.connect(url, { useNewUrlParser: true }, function (err, db) {
    if (err) throw err;

    var db_write = db.db("TestServer");

    var sensor_mac = "D2C91B20";
    var UID = 10015;
    var Name = "王小明"


    //1553240835494
    //1553240985105.09
    //149611 ms

    async.forever(function (callback) {

        if (err) throw err;
        //var start = Number(Date.now());
        //console.log("Start time");
        //console.log(start);
        fs.createReadStream('ecg_data_50000.csv')
            .pipe(csv())
            .on('data', (row) => {
                //console.log("data");

                var current_time = Number(Date.now());
                var start_time = 1561443546112;
                var t_shift = current_time - start_time;
                var t = Number(Math.floor(row.timestamp)) + t_shift;
                var ecg_data = [
                    { "CH": "L1", "value": Number(row.L1) },
                    { "CH": "L2", "value": Number(row.L2) },
                    { "CH": "L3", "value": Number(row.L3) },
                    { "CH": "aVL", "value": Number(row.aVL) },
                    { "CH": "aVF", "value": Number(row.aVF) },
                    { "CH": "aVR", "value": Number(row.aVR) },
                    { "CH": "v1", "value": Number(row.v1) },
                    { "CH": "v2", "value": Number(row.v2) },
                    { "CH": "v3", "value": Number(row.v3) },
                    { "CH": "v4", "value": Number(row.v4) },
                    { "CH": "v5", "value": Number(row.v5) },
                    { "CH": "v6", "value": Number(row.v6) }
                ];

                var mongo_obj = {
                    UID: UID,
                    Name: Name,
                    Sensor_mac: sensor_mac,
                    ECG_data: ecg_data,
                    Timestamp: t
                };

                db_write.collection("ecg").insertOne(mongo_obj);


            })
            .on('end', () => {
                //console.log('CSV file successfully processed');
                //var end = Date.now();
                //console.log(end);
                setTimeout(function () {
                    callback();
                }, 197332) // usdrId = 1 , data length = 197332 ms , id >= 8897527 , <= 8947526
            });
    }, function (err) {
        //console.log(err);
    });

});

/*
{
    "UID": 10035,
    "Name": "王小明",
    "Sensor_mac":"D2C91B20",
    "ECG_data":[
        {"CH" : "L1", "value" : 0.1546},
        {"CH" : "L2", "value" : 0.0413},
        {"CH" : "L3", "value" : 0.1784},
        {"CH" : "aVL", "value" : 0.3361},
        {"CH" : "aVF", "value" : 0.5361},
        {"CH" : "aVR", "value" : 0.1219},
        {"CH" : "v1", "value" : -0.0354},
        {"CH" : "v2", "value" : -0.1278},
        {"CH" : "v3", "value" : 0.19237},
        {"CH" : "v4", "value" : 0.7123},
        {"CH" : "v5", "value" : -0.0069},
        {"CH" : "v6", "value" : 0.01701}
    ],
    "Timestamp" : 1566202810413
}
*/

File: InsertCSV_data.py
import time
import pymongo
import csv
client = pymongo.MongoClient(host = 'localhost',port=27017)
db = client.TestServer
collection = db.ecg


while True:

    with open('ecg_data_50000.csv',newline = '') as csvfile:
        current_time = time.time_ns()
        print(current_time)
        t_shift = int(current_time/1000000 - 1561443546112)
        print(t_shift)
        
        rows = csv.reader(csvfile)
        ind =0
        for row in rows:
            if ind !=0:
                mongo_obj = {
                        'Device_ID' :'AABBCCDD0011',
                        'BLE_Name': 'ECG-XXXXXXXX',
                        'ECG_data': [
                                {"CH":"L1",'value':row[2]},
                                {"CH":"L2",'value':row[3]},
                                {"CH":"L3",'value':row[4]},
                                {"CH":"V1",'value':row[5]},
                                {"CH":"V2",'value':row[6]},
                                {"CH":"V3",'value':row[7]},
                                {"CH":"V4",'value':row[8]},
                                {"CH":"V5",'value':row[9]},
                                {"CH":"V6",'value':row[10]},
                                {"CH":"aVR",'value':row[11]},
                                {"CH":"aVL",'value':row[12]},
                                {"CH":"aVF",'value':row[13]}
                                ],
                        'Timestamp': int(float(row[1])) + int(t_shift) + 8*60*60
                        }

            
            if ind !=0:
                collection.insert_one(mongo_obj)
            if ind == 0:
                ind = 1
        end_time = time.time_ns() 
        print(end_time)
        print((end_time-current_time)/1000000)
        time.sleep((197332-(end_time-current_time)/1000000)/1000)
'''
        current_time = time.time_ns()
        print(current_time)
        t_shift = int(current_time/1000000 - 1561443546112)
        print(t_shift)
        ind = 0
        end_time = time.time_ns() 
        print(end_time)
        print((end_time-current_time)/1000000)
        time.sleep(1000)
'''        

        
        
'''    
    
        for row in rows:
            if ind !=0:
                mongo_obj = {
                        'UID' :10015,
                        'Name': 'Test',
                        'Sensor_mac': 'B1C73A25',
                        'ECG_data': [
                                {"CH":"L1",'value':row[2]},
                                {"CH":"L2",'value':row[3]},
                                {"CH":"L3",'value':row[4]},
                                {"CH":"V1",'value':row[5]},
                                {"CH":"V2",'value':row[6]},
                                {"CH":"V3",'value':row[7]},
                                {"CH":"V4",'value':row[8]},
                                {"CH":"V5",'value':row[9]},
                                {"CH":"V6",'value':row[10]},
                                {"CH":"aVR",'value':row[11]},
                                {"CH":"aVL",'value':row[12]},
                                {"CH":"aVF",'value':row[13]}
                                ],
                        'Timestamp': int(float(row[1])) + int(t_shift/1000) + 8*60*60
                        }
            
            if ind !=0:
                collection.insert_one(mongo_obj)
            if ind == 0:
                ind = 1
'''    

    
        #time.sleep(197332-(end_time-current_time)/1000000)
        
#client.close()
    
    


